liveness: allow registering callbacks after start

tbg opened the pull request:
I discovered[^1] a deadlock scenario when multiple nodes in the cluster restart
with additional stores that need to be bootstrapped. In that case, liveness
must be running when the StoreIDs are allocated, but it is not.

Trying to address this problem, I realized that when an auxiliary Store is bootstrapped,
it will create a new replicateQueue, which will register a new callback into NodeLiveness.

But if liveness must be started at this point to fix #106706, we'll run into the assertion
that checks that we don't register callbacks on a started node liveness.

Something's got to give: we will allow registering callbacks at any given point
in time, and they'll get an initial set of notifications synchronously. I
audited the few users of RegisterCallback and this seems OK with all of them.

[^1]: https://github.com/cockroachdb/cockroach/issues/106706#issuecomment-1640254715

Epic: None
Release note (bug fix): it was possible for a node status to reflect a "last up" timestamp that lead the actual last liveness heartbeat of the node. This has been fixed.

tbg committed:
liveness: allow registering callbacks after start

I discovered[^1] a deadlock scenario when multiple nodes in the cluster restart
with additional stores that need to be bootstrapped. In that case, liveness
must be running when the StoreIDs are allocated, but it is not.

Trying to address this problem, I realized that when an auxiliary Store is bootstrapped,
it will create a new replicateQueue, which will register a new callback into NodeLiveness.

But if liveness must be started at this point to fix #106706, we'll run into the assertion
that checks that we don't register callbacks on a started node liveness.

Something's got to give: we will allow registering callbacks at any given point
in time, and they'll get an initial set of notifications synchronously. I
audited the few users of RegisterCallback and this seems OK with all of them.

[^1]: https://github.com/cockroachdb/cockroach/issues/106706#issuecomment-1640254715

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107265)
<!-- Reviewable:end -->

erikgrinaker commented:
Right, this is probably fallout from #103601.

andrewbaptist commented:
Thanks Tobi!

I see what you mean about this test, and it was tricky to get some of tests, including this one, to work without doing what you did by allowing later registration. 

The approach you have works, but I had considered two other options as well (and could code them up if you want)
1) Don't allow disks/stores to be added to a running instance, instead require a restart. I wasn't sure if this was something that we support in a field deployment but I should have checked first. If we don't support this, then it would be possible to rearrange the code to allow this to work

2) Have a "single" node level callback that then fans out to the stores on liveness changes. This simplifies the liveness code as there is only a single callback, and the node already has locking in place for adding stores.

Your approach also works and is probably the most straightforward.

tbg commented:
> Don't allow disks/stores to be added to a running instance, instead require a restart. I wasn't sure if this was something that we support in a field deployment but I should have checked first. If we don't support this, then it would be possible to rearrange the code to allow this to work

The new stores are being added following a restart. However, due to the need to assign StoreIDs, KV needs to be up and running. So these new stores are added "async". I don't see a good way around that.

> Have a "single" node level callback that then fans out to the stores on liveness changes. This simplifies the liveness code as there is only a single callback, and the node already has locking in place for adding stores.

That's not a bad idea, though I won't be able to do this until I leave (and it might be crowded out anyway). So even though I like that idea I'd like to stick with what I have, so that we can at least address the bug.

erikgrinaker commented on pkg/kv/kvserver/liveness/liveness.go:
```diff
@@ -641,13 +647,7 @@ func (nl *NodeLiveness) Start(ctx context.Context) {
-			}
-		}
-	}
+	nl.notifyIsAliveCallbacks()
```
This doesn't actually notify anything, it needs to pass in the callbacks, or the method needs to collect them. This also reveals lacking test coverage, can we add a quick test?

tbg replied:
It's now tested, and I removed the footgun (need to pass a `[]IsLiveCallback`). This required some refactoring, basically picking up where @andrewbaptist left off at some point in the past (thanks for the initial leg work!)

tbg committed:
liveness: only call onSelfHeartbeat on self

I think there was a bug here. This method was previously invoked in
`updateLiveness`, but that method is the general workhorse for updating
anyone's liveness. In particular, it is called by `IncrementEpoch`. So we were
invoking `onSelfHeartbeat` when we would increment other nodes' epochs. This
doesn't seem great.

Additionally, the code was trying to avoid invoking this callback before
liveness was officially "started". Heartbeating yourself before liveness is
started is unfortunately a thing due to the tangled start-up initialization
sequence; we may see heartbeats triggered by lease requests.

Avoid both complications by invoking `onSelfCallback` from the actual main
heartbeat loop, whose only job is to heartbeat the own liveness record.

I tried to adopt `TestNodeHeartbeatCallback` to give better coverage, but
it's a yak shave. A deterministic node liveness (i.e. a way to invoke the
main heartbeat loop manually) would make this a lot simpler. I filed an
issue to that effect:

https://github.com/cockroachdb/cockroach/issues/107452

tbg committed:
gossip: add GetNodeID accessor

tbg committed:
liveness: add and adopt Gossip interface

This will help with testing.

tbg committed:
liveness: rename storage -> storageImpl

We'll give this a proper interface soon.

tbg committed:
liveness: export storage methods

So that it can implement a public interface.

tbg committed:
liveness: switch storage to ptr receivers

tbg committed:
liveness: half-adopt Storage

We still want a Storage to be passed into NewNodeLiveness as opposed to a
`*kv.DB`, but so far, so good.

tbg committed:
liveness: finish adopting Storage

Now Liveness is constructed using a `Storage` as opposed to a `*kv.DB`.

tbg committed:
liveness: move liveness and store regex to globals

Helps with testing.

tbg committed:
liveness: add test for IsLiveCallback invocation

This tests that regardless of when a callback is registered, it gets called.

tbg committed:
liveness: export LivenessUpdate

It's needed to implement the liveness storage (once it exists).

tbg committed:
kvserver: avoid implicit nils in TestReplicaLeaseCounters

The two removed fields are nil. This made a test failure during the refactors
in this PR more annoying. If we're going to set up a half-inited NodeLiveness,
let's at least be honest about it.

erikgrinaker commented on pkg/kv/kvserver/liveness/liveness.go:
```diff
@@ -700,6 +687,8 @@ func (nl *NodeLiveness) Start(ctx context.Context) {
 				}); err != nil {
 				log.Warningf(ctx, heartbeatFailureLogFormat, err)
+			} else if nl.onSelfHeartbeat != nil {
+				nl.onSelfHeartbeat(ctx)
```
This misses a bunch of cases. In particular, explicit `Heartbeat()` calls made during lease acquisitions.

We may want to move this into `heartbeatInternal` or something, or keep it in `updateLiveness` with the proper condition.

tbg replied:
> This misses a bunch of cases. In particular, explicit Heartbeat() calls made during lease acquisitions.

This was intentional, see commit message. There is no need for "every" heartbeat to bump this timestamp. But we do want the last up timestamp to rise along at the pace of the heartbeat loop, which this code achieves.

The last up timestamp is used only to inform the "node up" event that is emitted to the event log. There is no way to record the "true" last time the node is up anyway (what does that even mean?), so this mechanism keeps it simple and gets the job done.

Please let me know if you think I'm missing something here.

andrewbaptist replied:
This is definitely a bug before. The `onSelfHeartbeat` call should not be made for other nodes. Thanks for catching and fixing it!

erikgrinaker replied:
> The last up timestamp is used only to inform the "node up" event that is emitted to the event log.

Oh right, that thing. I'd be inclined to just remove that outright, but yeah, let's leave that for another day.

andrewbaptist reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

erikgrinaker commented on pkg/kv/kvserver/liveness/storage.go:
```diff
@@ -24,15 +24,32 @@ import (
+}
+
+// storageImpl implements Storage by storing entries in the replicated KV liveness range.
+type storageImpl struct {
```
nit: let's rename this `kvStorage`, c.f. `NewKVStorage`.

erikgrinaker commented:
We'll want a targeted backport here too, since it seems like prior versions are vulnerable to the same issue (but I'm not sure with all the refactoring in this area).

tbg commented:
> We'll want a targeted backport here too, since it seems like prior versions are vulnerable to the same issue (but I'm not sure with all the refactoring in this area).

I'm not sure we should fix this in a backport. The chances of messing something up are much higher than anyone hitting this bug. Folks very very rarely add stores to existing nodes (multi-storage in itself is rare) and doing so in a way that loses quorum on the storeIDGen is even rarer. Also, there is a workaround - restart a quorum without the extra store.

If you disagree, let's file an issue because I am not sure I would get to this backport even if we tried.

erikgrinaker commented:
Ok, I can buy that. Was thinking of surrounding issues like the self callback too though, but they're fairly minor.

tbg commented:
The self-callback I can fix in a backport - we can just add a conditional (if we don't insist on also adding a test).

The hang at startup, btw, I think has existed in every past version of CRDB, and to the best of my knowledge was never hit by anyone in a real cluster. Even the test that caught it only did so after I refactored it and made it use in-mem engines. So I am fairly confident that _not_ backporting is the reasonable strategy here.

tbg commented:
bors r=erikgrinaker

tbg commented:
Backports:


https://github.com/cockroachdb/cockroach/pull/107605
https://github.com/cockroachdb/cockroach/pull/107606
